### ISSUE TITLE
Identify 'code smell' and refactor.

### ISSUE THREAD

**Issue by v0dro** [SciRuby/rubex#35]
Rubex was written in a bit of hurry, so parts of the code base are quite messy. Identify these smelly parts and refactor them with better design. Refer to Martin Fowler's Refactoring Ruby book for reference.

**Comment by v0dro** [comment]
Refactoring should be done in the refactor branch: https://github.com/SciRuby/rubex/tree/refactor

**Comment by shaunakpp** [comment]
Do we have a Roadmap or Milestones planned for refactoring? 
Also, what needs refactoring and what not can be subjective, so can we keep a dedicated channel for discussing this?

**Comment by v0dro** [comment]
Works. How do you think a channel should be setup? Gitter or something? 

I'm currently figuring out things to refactor myself. Will post my list in this thread now.

**Comment by v0dro** [comment]
I've updated with a rough list.

**Comment by shaunakpp** [comment]
Gitter or slack would be best.
Also, one thing I noticed was that files in `lib/rubex/ast/` implement multiple classes in one single file. If we keep separate files for separate classes, it will help in maintaining the project and also reduce the overall code-churn. What do you think?

**Comment by v0dro** [comment]
So its mainly done that way because all the classes under say `expression.rb` cater to only expressions, and most of these classes tend to be interdependent so it makes it easier to read and make edits between classes. Also, most of them are quite small too (average of 40 lines, max 100 lines) so I didn't think it would be best to split them into multiple files.

Is having a class per file a best practice, though?

I'll setup Gitter soon. I'll also open a #rubex channel on the sciruby slack.

**Comment by shaunakpp** [comment]
> Is having a class per file a best practice, though?

Yup. Check [this](https://github.com/bbatsov/ruby-style-guide#one-class-per-file)

**Comment by v0dro** [comment]
OK. I'm sold. I'm updating the above list to reflect this change.

**Comment by shaunakpp** [comment]
I can start working on separating the classes. If no-one else has already taken it up.

**Comment by v0dro** [comment]
Sure. Please commit your changes to the 'refactor' branch.

**Comment by v0dro** [comment]
MethodCall#analyse_statement: https://github.com/SciRuby/rubex/commit/0779f5660df554debd3ff0ddb210723c5976762e

**Comment by v0dro** [comment]
CommandCall#generate_evaluation_code : https://github.com/SciRuby/rubex/commit/012963f63c99ca767dc5567238a72775aaf80dce

**Comment by v0dro** [comment]
Name#analyse_statement: https://github.com/SciRuby/rubex/commit/a31144b4900da684b26892516289a866447ead88

**Comment by v0dro** [comment]
ElementRef: https://github.com/SciRuby/rubex/commit/583d85980bc6b9361fe2fa2206e9fc375e48805e

**Comment by v0dro** [comment]
@shaunakpp any progress on your end?

**Comment by shaunakpp** [comment]
I'll attempt it this weekend. Been a bit busy with work.

Sent from my Nokia 3310
________________________________
